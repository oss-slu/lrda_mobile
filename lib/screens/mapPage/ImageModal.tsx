import React, { useState } from 'react';
import {
  Modal,
  View,
  ScrollView,
  Text,
  Image,
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  Dimensions,
} from 'react-native';
import { useTheme } from "../../components/ThemeProvider";

// Assuming `Image` type was intended to be an array of objects with a 'uri' property
interface ImageType {
  uri: string;
}

interface Props {
  isVisible: boolean;
  onClose: () => void;
  images: ImageType[];
}

const { width, height } = Dimensions.get("window");

// Correctly destructure props and use them directly instead of internal state for visibility
const ImageModal: React.FC<Props> = ({ isVisible, onClose, images }) => {
  const [imageLoadedState, setImageLoadedState] = useState<{ [key: string]: boolean }>({});
  const [isImageTouched, setIsImageTouched] = useState(false);
  const theme = useTheme(); // Assuming `useTheme` returns the current theme

  const handleLoad = (uri: string) => {
    setImageLoadedState((prev) => ({ ...prev, [uri]: true }));
  };

  // Define a missing handler if needed, or remove if not used
  const handleImageTouchStart = () => setIsImageTouched(!isImageTouched);

  return (
    <Modal animationType="slide" transparent={false} visible={isVisible} onRequestClose={onClose}>
      <View style={styles.container}>
        <ScrollView
          style={{ height: isImageTouched ? "80%" : "50%" }}
          onTouchStart={images && images.length > 2 ? handleImageTouchStart : undefined}
        >
          {images && images.length > 0 ? (
            images.map((image, index) => (
              <View key={index} style={styles.imageContainer}>
                {!imageLoadedState[image.uri] && (
                  <ActivityIndicator size="large" color="#0000ff" />
                )}
                <Image
                  source={{ uri: image.uri }}
                  style={styles.image}
                  onLoad={() => handleLoad(image.uri)}
                />
              </View>
            ))
          ) : (
            <Text style={styles.noImagesText}>No images</Text>
          )}
        </ScrollView>

        <TouchableOpacity style={styles.closeButton} onPress={onClose}>
          <Text>Close</Text>
        </TouchableOpacity>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  imageContainer: {
    alignItems: 'center',
    width: '100%',
  },
  image: {
    width: width,
    height: width,
  },
  noImagesText: {
    alignSelf: 'center',
    justifyContent: 'center',
    marginTop: 200,
  },
  closeButton: {
    height: 40,
    width: 75,
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 20,
    backgroundColor: '#ddd',
    padding: 10,
    borderRadius: 5,
    marginBottom: 30,
  },
});

export default ImageModal;
