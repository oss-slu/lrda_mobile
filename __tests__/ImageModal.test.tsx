import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import ImageModal from '../lib/screens/mapPage/ImageModal';

// Mock ThemeProvider
jest.mock('../lib/components/ThemeProvider', () => ({
  useTheme: () => ({
    theme: {
      text: '#000',
    },
  }),
}));

beforeEach(() => {
  // Clear mocks before each test
  jest.clearAllMocks();

  // Mock console methods to avoid unnecessary log outputs in tests
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});
  jest.spyOn(console, "warn").mockImplementation(() =>{});
  jest.spyOn(console, 'warn').mockImplementation((message) => {
    if (!message.includes('Toolbar has no editor')) {
      console.warn(message);
    }
  });
});

afterEach(() => {
  // Restore the original console methods
  console.log.mockRestore();
  console.error.mockRestore();
  console.warn.mockRestore();
});

describe('ImageModal', () => {
  const mockImages = [
    { uri: 'https://example.com/image1.jpg' },
    { uri: 'https://example.com/image2.jpg' },
  ];

  it('renders without crashing', () => {
    const { toJSON } = render(<ImageModal isVisible={true} onClose={() => {}} images={mockImages} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays the correct number of images', () => {
    const { getAllByTestId } = render(<ImageModal isVisible={true} onClose={() => {}} images={mockImages} />);
    const images = getAllByTestId('image-component'); // Assuming each Image component has a testID 'image-component'
    expect(images.length).toBe(1);
  });

  it('handles the close button press', () => {
    const mockOnClose = jest.fn();
    const { getByTestId } = render(<ImageModal isVisible={true} onClose={mockOnClose} images={mockImages} />);
    const closeButton = getByTestId('close-button'); // Assuming the TouchableOpacity has testID 'close-button'
    
    fireEvent.press(closeButton);
    expect(mockOnClose).toHaveBeenCalled();
  });
});
