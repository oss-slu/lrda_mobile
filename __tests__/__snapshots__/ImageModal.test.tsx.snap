// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageModal renders without crashing 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": undefined,
        "flex": 1,
        "justifyContent": "center",
        "paddingTop": 45,
        "width": "100%",
      }
    }
  >
    <RCTScrollView
      style={
        {
          "height": "50%",
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": undefined,
              "width": "100%",
            }
          }
        >
          <ActivityIndicator
            color="#0000ff"
            size="large"
          />
          <Image
            onLoad={[Function]}
            source={
              {
                "uri": "https://example.com/image1.jpg",
              }
            }
            style={
              {
                "height": 750,
                "width": 750,
              }
            }
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": undefined,
              "width": "100%",
            }
          }
        >
          <ActivityIndicator
            color="#0000ff"
            size="large"
          />
          <Image
            onLoad={[Function]}
            source={
              {
                "uri": "https://example.com/image2.jpg",
              }
            }
            style={
              {
                "height": 750,
                "width": 750,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ddd",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 30,
          "marginTop": 20,
          "opacity": 1,
          "padding": 10,
          "width": 75,
        }
      }
      testID="image-component"
    >
      <Text
        style={
          {
            "fontFamily": "Inter",
          }
        }
        testID="close-button"
      >
        Close
      </Text>
    </View>
  </View>
</Modal>
`;
